### PR TITLE
[TEST] Fix compiling problem and removed -DENABLE_TEST

### LINKED ISSUE #2400: Can not compile using bazel without `--copt=-DENABLE_TEST`
**Describe your environment** 

OS: Linux
Compiler: GCC 9
Build system: bazel

**Steps to reproduce**

`bazel build ...  --copt=-DENABLE_TEST  --copt=-DENABLE_ASYNC_EXPORT`

**What is the expected behavior?**
Build success.

**What is the actual behavior?**

```bash
/data/home/workspace/testopentelemetry/opentelemetry-cpp/ext/test/http/BUILD:4:8: Compiling ext/test/http/curl_http_test.cc failed: (Exit 1): gcc failed: error executing command /usr/local/gcc-9.4.0/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 63 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
ext/test/http/curl_http_test.cc: In member function 'virtual void BasicCurlHttpTests_SendGetRequestAsync_Test::TestBody()':
ext/test/http/curl_http_test.cc:417:17: error: 'class opentelemetry::v1::ext::http::client::curl::HttpClient' has no member named 'WaitBackgroundThreadExit'
  417 |     http_client.WaitBackgroundThreadExit();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
ext/test/http/curl_http_test.cc: In member function 'virtual void BasicCurlHttpTests_SendPostRequestAsync_Test::TestBody()':
ext/test/http/curl_http_test.cc:490:17: error: 'class opentelemetry::v1::ext::http::client::curl::HttpClient' has no member named 'WaitBackgroundThreadExit'
  490 |     http_client.WaitBackgroundThreadExit();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
ext/test/http/curl_http_test.cc: In member function 'virtual void BasicCurlHttpTests_FinishInAsyncCallback_Test::TestBody()':
ext/test/http/curl_http_test.cc:519:17: error: 'class opentelemetry::v1::ext::http::client::curl::HttpClient' has no member named 'WaitBackgroundThreadExit'
  519 |     http_client.WaitBackgroundThreadExit();
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~
INFO: Elapsed time: 301.171s, Critical Path: 50.79s
INFO: 2683 processes: 371 internal, 2312 linux-sandbox.
FAILED: Build did NOT complete successfully
```

### PR DESCRIPTION
**Pull request by owent**
Fixes #2400

## Changes

+ Fix compiling problem and removed -DENABLE_TEST

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

**Review by marcalff — COMMENTED** [review]
Could the "ENABLE_TEST" logic be removed entirely instead ?

It is only used for http_client_curl.h, which makes it very suspicious.

How about replacing:

```
HttpClientFactory::CreateNoSend()
```

with

```
HttpClientTestFactory::Create()
```

File: exporters/otlp/test/otlp_http_exporter_test.cc [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #  include "opentelemetry/sdk/trace/batch_span_processor.h"
 #  include "opentelemetry/sdk/trace/batch_span_processor_options.h"
 #  include "opentelemetry/sdk/trace/tracer_provider.h"
+#  include "opentelemetry/test_common/ext/http/client/http_client_factory.h"
```
Comment (marcalff) [inline exporters/otlp/test/otlp_http_exporter_test.cc:22]: include http_client_test_factory.h instead ?

File: exporters/otlp/test/otlp_http_log_record_exporter_test.cc [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #  include "opentelemetry/sdk/logs/exporter.h"
 #  include "opentelemetry/sdk/logs/logger_provider.h"
 #  include "opentelemetry/sdk/resource/resource.h"
+#  include "opentelemetry/test_common/ext/http/client/http_client_factory.h"
```
Comment (marcalff) [inline exporters/otlp/test/otlp_http_log_record_exporter_test.cc:25]: include http_client_test_factory.h instead ?

File: exporters/otlp/test/otlp_http_metric_exporter_test.cc [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #include "opentelemetry/sdk/metrics/export/metric_producer.h"
 #include "opentelemetry/sdk/metrics/instruments.h"
 #include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/test_common/ext/http/client/http_client_factory.h"
```
Comment (marcalff) [inline exporters/otlp/test/otlp_http_metric_exporter_test.cc:27]: include http_client_test_factory.h instead ?

File: test_common/src/http/client/nosend/http_client_test_factory.cc [hunk 4]
```diff
@@ -1,15 +1,13 @@
 // Copyright The OpenTelemetry Authors
 // SPDX-License-Identifier: Apache-2.0
 
-#ifdef ENABLE_TEST
-#  include "opentelemetry/ext/http/client/http_client.h"
-#  include "opentelemetry/ext/http/client/http_client_factory.h"
-#  include "opentelemetry/test_common/ext/http/client/nosend/http_client_nosend.h"
+#include "opentelemetry/ext/http/client/http_client.h"
+#include "opentelemetry/test_common/ext/http/client/http_client_factory.h"
```
Comment (marcalff) [inline test_common/src/http/client/nosend/http_client_test_factory.cc:5]: include http_client_test_factory.h instead ?

**Review by marcalff — COMMENTED** [review]
A few build fixes

**Comment by owent** [comment]
> Could the "ENABLE_TEST" logic be removed entirely instead ?
> 
> It is only used for http_client_curl.h, which makes it very suspicious.
> 
> How about replacing:
> 
> ```
> HttpClientFactory::CreateNoSend()
> ```
> 
> with
> 
> ```
> HttpClientTestFactory::Create()
> ```

Good point. `-DENABLE_TEST` is removed.

**Review by marcalff — APPROVED** [review]
LGTM. Thanks for the cleanup.

**Review by esigo — COMMENTED** [review]
LGTM
Thanks  :)

**Review by esigo — APPROVED** [review]

**Pull request merged**